pkg/virt-handler/cache/cache.go

```diff
@@ -0,0 +1,197 @@
+/*
+ * This file is part of the KubeVirt project
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ * Copyright 2017 Red Hat, Inc.
+ *
+ */
+
+package cache
+
+import (
+	"sync"
+	"time"
+
+	k8sv1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/watch"
+	"k8s.io/client-go/tools/cache"
+
+	"kubevirt.io/kubevirt/pkg/log"
+	"kubevirt.io/kubevirt/pkg/virt-launcher/virtwrap/api"
+	cmdclient "kubevirt.io/kubevirt/pkg/virt-launcher/virtwrap/cmd-server/client"
+	notifyserver "kubevirt.io/kubevirt/pkg/virt-launcher/virtwrap/notify-server"
+	"kubevirt.io/kubevirt/pkg/watchdog"
+)
+
+func newListWatchFromNotify(virtShareDir string, watchdogTimeout int) cache.ListerWatcher {
+	d := &DomainWatcher{
+		backgroundWatcherStarted: false,
+		virtShareDir:             virtShareDir,
+		watchdogTimeout:          watchdogTimeout,
+	}
+
+	return d
+}
+
+type DomainWatcher struct {
+	lock                     sync.Mutex
+	wg                       sync.WaitGroup
+	stopChan                 chan struct{}
+	eventChan                chan watch.Event
+	backgroundWatcherStarted bool
+	virtShareDir             string
+	watchdogTimeout          int
+}
+
+func (d *DomainWatcher) startBackground() error {
+	d.lock.Lock()
+	defer d.lock.Unlock()
+
+	if d.backgroundWatcherStarted == true {
+		return nil
+	}
+
+	d.stopChan = make(chan struct{}, 1)
+	d.eventChan = make(chan watch.Event, 100)
+
+	d.wg.Add(1)
+	go func() {
+		defer d.wg.Done()
+
+		expiredWatchdogTicker := time.NewTicker(time.Duration(d.watchdogTimeout) * time.Second).C
+		srvErr := make(chan error)
+		go func() {
+			defer close(srvErr)
+			err := notifyserver.RunServer(d.virtShareDir, d.stopChan, d.eventChan)
```

Comment:
The whole notify server might also be better placed in the virt-handler package.

Reply:
i'm torn about that one. The notify client depends on libvirt. I could move the server to virt-handler but I'd want to keep the client package under virt-launcher.  i thought it was better to keep them both together in one place.

Reply:
fixed. I moved notify server and cmd client to virt-handler